flare/gp_from_aimd.py

```diff
@@ -0,0 +1,232 @@
+                train_atoms.append(atoms_of_specie[0])
+
+            self.update_gp_and_print(cur_frame, train_atoms, train=True)
+
+        if len(self.gp.training_data) > 0 and self.gp.l_mat is None:
+            self.gp.train(monitor=self.verbose)
+
+        # Loop through trajectory
+        for i, cur_frame in enumerate(self.frames):
+
+            if self.verbose >= 2:
+                print("=====NOW ON FRAME {}=====".format(i))
+            dft_forces = deepcopy(cur_frame.forces)
+            self.pred_func(cur_frame, self.gp)
+
+            # Convert to meV/A
+            mae = np.mean(np.abs(cur_frame.forces - dft_forces)) * 1000
+            mac = np.mean(np.abs(dft_forces)) * 1000
+
+            output.write_gp_dft_comparison(
+                curr_step=i, frame=cur_frame,
+                start_time=time.time(), output_name=self.output_name,
+                dft_forces=dft_forces,
+                mae=mae, mac=mac, local_energies=None)
+
+            # get max uncertainty atoms
+            std_in_bound, train_atoms = self.is_std_in_bound(cur_frame)
+            if not std_in_bound:
+
+                # compute mae and write to output
+                # add max uncertainty atoms to training set
+                self.update_gp_and_print(cur_frame, train_atoms, train=False)
+
+                if self.train_count < self.max_trains:
+                    self.train_gp()
+
+        output.conclude_run(self.output_name)
+
+        if self.pickle_name:
+            with open(self.pickle_name, 'wb') as f:
+                pickle.dump(self.gp, f)
+
+    def update_gp_and_print(self, frame, train_atoms: List[int], train=True):
+        """
+        Update the internal GP model training set with a list of training
+        atoms indexing atoms within the frame. If train is True, re-train
+        the GP by optimizing hyperparameters.
+        :param frame:
+        :param train_atoms:
+        :param train:
+        :return:
+        """
+
+        output.write_to_output('\nAdding atom(s) {} to the '
+                               'training set.\n'
+                               .format(train_atoms, ),
+                               self.output_name)
+        output.write_to_output('Uncertainties: {}.\n'
+                               .format(frame.stds[train_atoms]),
+                               self.output_name)
+
+        # update gp model
+        self.gp.update_db(frame, frame.forces, custom_range=train_atoms)
+        self.gp.set_L_alpha()
+        # TODO double check that these are being called at the right time.
+        if train:
+            self.train_gp()
+
+    def train_gp(self):
+        self.gp.train(monitor=True if self.verbose >= 2 else False)
+
+        output.write_hyps(self.gp.hyp_labels, self.gp.hyps,
+                          self.start_time, self.output_name,
+                          self.gp.like, self.gp.like_grad)
+        self.train_count += 1
+
+    def is_std_in_bound(self, frame):
```

Comment:
Looks similar to the is_std_in_bound method of otf. To avoid code duplication, we should have it as a function in one place (either otf or a separate file) and import from there.

Reply:
Agreed. I can transfer this to a different file and re-factor OTF to point there (note that this implementation borrows the good idea of @nw13slx to allow a relative and absolute cutoff simultaneously, and I unit-tested every permutation). I think a good home could be within  predict.py, as the functionality it serves is to be a diagnostic to a prediction made on a given atom.